src/source-maps.cc

```diff
@@ -0,0 +1,187 @@
+      const auto& seg = group.segments[j];
+      const Segment* last_seg = nullptr;
+      if (j > 0) last_seg = &group.segments[j - 1];
+      if (seg.generated_col_delta == 0) INVALID();
+      if (!seg.has_source && seg.has_name) INVALID();
+      if (!seg.has_source) return true;
+      if (seg.source >= sources.size()) INVALID();
+      if (last_seg) {
+        if (seg.source_line_delta == 0 && seg.source_col_delta == 0) INVALID();
+        // FIXME: This has a limitation that if this seg has a source, the last
+        // one must.
+        if (last_seg->source_line + seg.source_line_delta != seg.source_line) {
+          INVALID();
+        }
+        if (last_seg->source_col + seg.source_col_delta != seg.source_col) {
+          INVALID();
+        }
+      }
+      if (!seg.has_name) return true;
+      if (seg.name >= names.size()) INVALID();
+    }
+  }
+  return true;
+}
+
+static int32_t cmpLocation(const SourceMapGenerator::SourceLocation& a,
+                           const SourceMapGenerator::SourceLocation& b) {
+  int32_t cmp = a.line - b.line;
+  return cmp ? cmp : a.col - b.col;
+}
+
+bool SourceMapGenerator::SourceMapping::operator<(
+    const SourceMapGenerator::SourceMapping& rhs) const {
+  int32_t cmp = cmpLocation(generated, rhs.generated);
+  if (cmp != 0) return cmp < 0;
+  cmp = cmpLocation(original, rhs.original);
+  if (cmp != 0) return cmp < 0;
+  if (source_idx == INDEX_NONE) return rhs.source_idx != INDEX_NONE;
+  if (rhs.source_idx == INDEX_NONE) return false;
+  return source_idx < rhs.source_idx;
+}
+
+bool SourceMapGenerator::SourceMapping::operator==(
+    const SourceMapGenerator::SourceMapping& rhs) const {
+  return !cmpLocation(generated, rhs.generated) &&
+         !cmpLocation(original, rhs.original) && source_idx == rhs.source_idx;
+}
+
+void SourceMapGenerator::SourceMapping::Dump() const {
+  std::cout << "Mapping " << original.line << ":" << original.col << " -> "
+            << generated.line << ":" << generated.col << " in " << source_idx
+            << "\n";
+}
+
+bool SourceMapGenerator::AddMapping(SourceLocation generated,
+                                    SourceLocation original,
+                                    std::string source) {
+  // Validate. For now, original, generated, and source are required
+  if (generated.line == 0 || original.line == 0) return false;
+  map_prepared = false;  // New mapping invalidates compressed map.
+  size_t source_idx = INDEX_NONE;
+  auto s = sources_map.find(source);
+  if (s == sources_map.end()) {
+    source_idx = map.sources.size();
+    map.sources.push_back(source);
+    bool inserted;
+    std::tie(std::ignore, inserted) = sources_map.insert({source, source_idx});
+    assert(inserted);
+  } else {
+    source_idx = s->second;
+  }
+  mappings.push_back({original, generated, source_idx});
+  return true;
+}
+
+void SourceMapGenerator::CompressMappings() {
+  std::sort(mappings.begin(), mappings.end());
```

Comment:
It's convenient for being self contained, but how likely are we to add mappings out of order? Seems like a lot of this could be simplified if we could assume that AddMapping only can add mappings in order.